src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js

```diff
@@ -73,6 +74,33 @@ angular.module('esn.inbox.libs')
       return mailbox;
     }
 
+    function mailboxtoTree(arr) {
+      const arrMap = new Map(arr.map(item => [item.id, item]));
+      const tree = [];
+
+      for (let i = 0, l = arr.length; i < l; i++) {
+        const item = arr[i];
+
+        if (item.parentId) {
+          const parentItem = arrMap.get(item.parentId);
+
+          if (parentItem && parentItem.id !== item.id) {
+            const { nodes } = parentItem;
+
+            if (nodes) {
+              parentItem.nodes.push(item);
+            } else {
+              parentItem.nodes = [item];
+            }
+          }
+        } else {
+          tree.push(item);
```

Comment:
what if the original array `arr` is not perfectly sorted?
example:

arr = [ parent, parent, parent, child, child, child ];

what happens is we will push the parents without linking them to the children.

so the tree will be [ parent ( no child ), parent ( no child ), parent ( no child ) ]

I think maybe we need to first sort the original array so that we ensure the child items are first to be processed and parents last ( for the mutation to work ).

Reply:
I think we don't need it's already done by James ;)